arangod/RocksDBEngine/RocksDBPersistedLog.h

```diff
@@ -49,17 +50,37 @@ struct RocksDBLogPersistor : std::enable_shared_from_this<RocksDBLogPersistor> {
   auto persist(std::shared_ptr<arangodb::replication2::replicated_log::PersistedLog> log,
                std::unique_ptr<arangodb::replication2::PersistedLogIterator> iter,
                WriteOptions const& options) -> futures::Future<Result>;
+  auto persist(std::shared_ptr<arangodb::replication2::replicated_log::PersistedLog> log,
+               replication2::LogIndex stop, WriteOptions const& options)
+      -> futures::Future<Result>;
 
   struct PersistRequest {
     PersistRequest(std::shared_ptr<arangodb::replication2::replicated_log::PersistedLog> log,
                    std::unique_ptr<arangodb::replication2::PersistedLogIterator> iter,
                    futures::Promise<Result> promise)
-        : log(std::move(log)), iter(std::move(iter)), promise(std::move(promise)) {}
+        : log(std::move(log)), action(InsertEntries{std::move(iter)}), promise(std::move(promise)) {}
+    PersistRequest(std::shared_ptr<arangodb::replication2::replicated_log::PersistedLog> log,
+                   replication2::LogIndex stop,
+                   futures::Promise<Result> promise)
+        : log(std::move(log)), action(RemoveFront{stop}), promise(std::move(promise)) {}
+
+    struct InsertEntries {
+      std::unique_ptr<arangodb::replication2::PersistedLogIterator> iter;
+    };
+
+    struct RemoveFront {
+      replication2::LogIndex stop;
+    };
+
+    auto execute(rocksdb::WriteBatch& wb) -> Result;
+
     std::shared_ptr<arangodb::replication2::replicated_log::PersistedLog> log;
-    std::unique_ptr<arangodb::replication2::PersistedLogIterator> iter;
+    std::variant<InsertEntries, RemoveFront> action;
     futures::Promise<Result> promise;
```

Comment:
Can these be const?

Reply:
If those member become `const` they have to copied during a move operation. `futures::Promise` is move-only. Additionally `Promise::setValue` is not `const`. `InsertEntries` contains a `unique_ptr` and is therefore move-only, hence the variant becomes move-only as well...

`log` can be const, but then again, it will be copied during move operations.

Reply:
Good point...